### PR TITLE
feat(nft-tags): add nft-tags

### PR DESCRIPTION
**Pull request by Daryl-L**
https://github.com/Magickbase/ckb-explorer-public-issues/issues/647

https://ckb-explorer-frontend-in-magickbase-repo-git-d27caf-magickbase.vercel.app/nft-collections

Because of using the component `MultiFilterButton` defined in https://github.com/Magickbase/ckb-explorer-frontend/pull/378, so this branch is based on it.

image

image

image

image

image

image

### CODE REVIEW

File: src/components/FilterSortContainer/index.tsx [hunk 1]
```diff
@@ -1,11 +1,24 @@
...
+    (acc, child, index) => {
+      if (index % 2 === 0) {
+        acc.left.push(child)
+      } else {
+        acc.right.push(child)
+      }
+      return acc
+    },
+    { left: [], right: [] },
+  )
```
Comment (Keith-CY) [inline src/components/FilterSortContainer/index.tsx:16]: CSS selector `odd | even` can be used if this is just for layout

Ref: https://developer.mozilla.org/en-US/docs/Web/CSS/:nth-child#keyword_values

File: src/pages/NftCollectionInfo/NftCollectionOverview/index.tsx [hunk 2]
```diff
@@ -7,6 +7,7 @@ import { explorerService } from '../../../services/ExplorerService'
 import styles from './styles.module.scss'
 import { handleNftImgError, patchMibaoImg } from '../../../utils/util'
 import { getPrimaryColor } from '../../../constants/common'
+import XUDTTag from '../../../components/XUDTTag'
```
Comment (Keith-CY) [inline src/pages/NftCollectionInfo/NftCollectionOverview/index.tsx:10]: Why use `xudt` tag in `nft` component

File: src/pages/NftCollections/List.tsx [hunk 3]
```diff
@@ -237,6 +294,13 @@ export const ListOnDesktop: React.FC<{ isLoading: boolean; list: NFTCollection[]
                     </Link>
                   </div>
                 </td>
+                <td>
+                  <div className={styles.tags}>
+                    {item.tags.map(tag => (
+                      <XUDTTag tagName={tag} />
```
Comment (Keith-CY) [inline src/pages/NftCollections/List.tsx:300]: `key` field is required for array of components

File: src/pages/NftCollectionInfo/NftCollectionOverview/index.tsx (hunk 2 above)
Comment (Daryl-L) [reply src/pages/NftCollectionInfo/NftCollectionOverview/index.tsx:10]: > Why use `xudt` tag in `nft` component

They are using the same style, so I just reuse it

Comment (Keith-CY) [reply src/pages/NftCollectionInfo/NftCollectionOverview/index.tsx:10]: They are semantically different so it's better not to reuse the code even now they are the same.

**Comment by Keith-CY** [comment]
Any update on the conversations above?

**Comment by Daryl-L** [comment]
> Any update on the conversations above?

solved

File: src/pages/NftCollections/index.tsx [hunk 4]
```diff
@@ -60,8 +61,11 @@ const NftCollections = () => {
           </a>
         </div>
         <div className={styles.list}>
-          <ListOnDesktop isLoading={isLoading} list={list} />
-          <ListOnMobile isLoading={isLoading} list={list} />
+          {isMobile ? (
+            <ListOnMobile isLoading={isLoading} list={tags === '' ? [] : list} />
```
Comment (Keith-CY) [inline src/pages/NftCollections/index.tsx:65]: Unnecessary `tags === ''`. If it is necessary, the condition can be prepended in

```js
const list = data?.data ?? []
```

at line 35

**Review by Keith-CY — APPROVED** [review]

File: src/pages/NftCollections/index.tsx (hunk 4 above)
Comment (Daryl-L) [reply src/pages/NftCollections/index.tsx:65]: > Unnecessary `tags === ''`. If it is necessary, the condition can be prepended in
> 
> ```js
> const list = data?.data ?? []
> ```
> 
> at line 35

`undefined` means query data without filter, and the empty string means not to query data, and  list nothing

Comment (Keith-CY) [reply src/pages/NftCollections/index.tsx:65]: > > Unnecessary `tags === ''`. If it is necessary, the condition can be prepended in
> > ```js
> > const list = data?.data ?? []
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > at line 35
> 
> `undefined` means query data without filter, and the empty string means not to query data, and list nothing

The statement is
```
if tags === '' then
  list is empty
else
  list is fetched from the network
```

That means the network request is unnecessary when `tags === ''`

Comment (Daryl-L) [reply src/pages/NftCollections/index.tsx:65]: Well, I think it is the same as what I did, if the tags is the empty string, use the empty array, or use the data from the api.

**Comment by FrederLu** [comment]
image
image

1、In the screenshot, the three tags cannot filter data. Are they not needed in the NFT Collection? At the same time, I see that there is an `RGB++ compatible` tag on the page, but there is no option for filtering.  cc @Sven-TBD 

---
image

2、When there is no data, paging needs to be handled.

---
image
image

3、After adding Tags to `NFT Collection`, the original `Created Time` and `Creator` styles became abnormal.

---
image

4、The font color in the label can be modified by referring to xudt.

File: src/pages/NftCollections/index.tsx (hunk 4 above)
Comment (Keith-CY) [reply src/pages/NftCollections/index.tsx:65]: Currently, the `list` will be fetched via HTTP request when `tags === ''`, but it's not used.

The steps are

1. fetch the list from backend
2. pass `[]` to the component when `tags === ''`

Step 1 can be avoided when `tags === ''`

**Comment by FrederLu** [comment]
>3、After adding Tags to NFT Collection, the original Created Time and Creator styles became abnormal.  

This problem still exists. @Daryl-L 

image

Because there are filter icons for both Tags and Type on NFT Collection, the positions are kept consistent. Now I find that the Tags label is a little too high.

**Comment by Daryl-L** [comment]
> > 3、After adding Tags to NFT Collection, the original Created Time and Creator styles became abnormal.
> 
> This problem still exists. @Daryl-L
> 
> image
> Because there are filter icons for both Tags and Type on NFT Collection, the positions are kept consistent. Now I find that the Tags label is a little too high.

There is no enough place to display the Create Time in one line when you reduce the width.

**Comment by FrederLu** [comment]
> > > 3、After adding Tags to NFT Collection, the original Created Time and Creator styles became abnormal.
> > 
> > 
> > This problem still exists. @Daryl-L
> > image
> > Because there are filter icons for both Tags and Type on NFT Collection, the positions are kept consistent. Now I find that the Tags label is a little too high.
> 
> There is no enough place to display the Create Time in one line when you reduce the width.

image

In the browser with 100% resolution, the styles of `Created Time` and `Creator` are abnormal. Can you try to shorten the distance between `Type` and `24hr Transactions` to make the last two items display normally?

File: src/pages/NftCollections/index.tsx (hunk 4 above)
Comment (Keith-CY) [reply src/pages/NftCollections/index.tsx:65]: Any update on this conversation?

**Comment by FrederLu** [comment]
https://ckb-explorer-frontend-in-magickbase-repo-rn19usuc5-magickbase.vercel.app/nft-collections
image
Why does this PC style look like a mobile version?  @Daryl-L

**Comment by FrederLu** [comment]
> > > > 3、After adding Tags to NFT Collection, the original Created Time and Creator styles became abnormal.
> > > 
> > > 
> > > This problem still exists. @Daryl-L
> > > image
> > > Because there are filter icons for both Tags and Type on NFT Collection, the positions are kept consistent. Now I find that the Tags label is a little too high.
> > 
> > 
> > There is no enough place to display the Create Time in one line when you reduce the width.
> 
> image
> In the browser with 100% resolution, the styles of `Created Time` and `Creator` are abnormal. Can you try to shorten the distance between `Type` and `24hr Transactions` to make the last two items display normally?

This problem still exists, please confirm again.

**Comment by Daryl-L** [comment]
> Visit Preview

Maybe the preview link has been updated
image

**Comment by FrederLu** [comment]
> > Visit Preview
> 
> Maybe the preview link has been updated image

image

Yes, this section has been updated in the new preveiw.

**Comment by FrederLu** [comment]
https://ckb-explorer-frontend-in-magickbase-repo-2v4r0h8do-magickbase.vercel.app/nft-collections?tags=out-of-length-range
image

1、When there is no data, the prompt text needs to be displayed in the center.

---
image
image

2、`Duplicate` tags are not supported on the NFT Api.

**Comment by Daryl-L** [comment]
> Magickbase/ckb-explorer-public-issues#647

base on https://vs0cjf.axshare.com/?id=tibwdv&p=_647___nft_tag_prd&g=1

**Comment by Keith-CY** [comment]
> @zmcNotafraid Could we adjust the max limit in the defination out-of-length-range to like 30 characters so that it could be more practical? cc @Keith-CY

Sure, the convention(https://talk.nervos.org/t/xudt-information-convention-rfc/8030/2) mentioned before has expired, so we can define our own rules.

**Review by FrederLu — APPROVED** [review]

**Pull request merged**